applications/main/nfc/plugins/supported_cards/bip.c

```diff
@@ -0,0 +1,372 @@
+            {
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+            },
+        .charges =
+            {
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+                {{0, 0, 0, 0, 0, 0}, 0},
+            },
+    };
+
+    const MfClassicData* data = nfc_device_get_data(device, NfcProtocolMfClassic);
+
+    do {
+        // verify all sectors' keys
+        for(size_t i = 0; i < mf_classic_get_total_sectors_num(data->type); i++) {
+            MfClassicSectorTrailer* sec_tr = mf_classic_get_sector_trailer_by_sector(data, i);
+            uint64_t key = bit_lib_bytes_to_num_be(sec_tr->key_a.data, 6);
+            if(key != bip_keys_a[i]) {
+                parsed = false;
+                break;
+            }
+            key = bit_lib_bytes_to_num_be(sec_tr->key_b.data, 6);
+            if(key != bip_keys_b[i]) {
+                parsed = false;
+                break;
+            }
+        }
+
+        // Get Card ID, little-endian 4 bytes at sector 0 block 1, bytes 4-7
+        parse_uint32_le(&data->block[SECTOR_BLOCK_OFFSET(0, 1)].data[4], &bip_data.card_id);
+
+        // Get balance, little-endian 2 bytes at sector 8 block 1, bytes 0-1
+        parse_uint16_le(&data->block[SECTOR_BLOCK_OFFSET(8, 1)].data[0], &bip_data.balance);
+
+        // Get balance flags (negative balance, etc.), little-endian 2 bytes at sector 8 block 1, bytes 2-3
+        parse_uint16_le(&data->block[SECTOR_BLOCK_OFFSET(8, 1)].data[2], &bip_data.flags);
+
+        // Get trip time window, proprietary format, at sector 5 block 1, bytes 0-7
+        parse_bip_timestamp(&data->block[SECTOR_BLOCK_OFFSET(5, 1)], &bip_data.trip_time_window);
+
+        // Last 3 top-ups: sector 10, ring-buffer of 3 blocks, timestamp in bytes 0-7, amount in bytes 9-10
+        for(size_t i = 0; i < 3; i++) {
+            if(is_bip_block_empty(&data->block[SECTOR_BLOCK_OFFSET(10, i)])) {
+                continue;
+            }
+            BipTransaction* top_up = &bip_data.top_ups[i];
+            parse_bip_timestamp(&data->block[SECTOR_BLOCK_OFFSET(10, i)], &top_up->timestamp);
+            parse_uint16_txn_amount(
+                &data->block[SECTOR_BLOCK_OFFSET(10, i)].data[9], &top_up->amount);
+        }
+
+        // Last 3 charges (i.e. trips), sector 11, ring-buffer of 3 blocks, timestamp in bytes 0-7, amount in bytes 10-11
+        for(size_t i = 0; i < 3; i++) {
+            if(is_bip_block_empty(&data->block[SECTOR_BLOCK_OFFSET(11, i)])) {
+                continue;
+            }
+            BipTransaction* charge = &bip_data.charges[i];
+            parse_bip_timestamp(&data->block[SECTOR_BLOCK_OFFSET(11, i)], &charge->timestamp);
+            parse_uint16_txn_amount(
+                &data->block[SECTOR_BLOCK_OFFSET(11, i)].data[10], &charge->amount);
+        }
+
+        parsed = true;
+    } while(false);
+
+    if(parsed) {
```

Comment:
Could you please move all logic below to the `do {} while(false)` structure above? I see that if we have parsed == true here, there is no chance that plugin wouldn't parse anything.

You can just move all code below to `do{} while(false)` structure above, leave `parse = true` at the end and return parsed value: `return parsed`

Reply:
Done! Would you like keeping it as it is now, where we first unpack everything into the `bip_data` struct and only print at the end? Or would you rather print as we unpack each bit of data?

Reply:
It's fine how it's done now. Thanks